pkg/apiserver/statement/statement.go

```diff
@@ -210,3 +228,118 @@ func (s *Service) getPlanDetailHandler(c *gin.Context) {
+// @Failure 401 {object} utils.APIError "Unauthorized failure"
+func (s *Service) downloadTokenHandler(c *gin.Context) {
+	var req GetStatementsRequest
+	if err := c.ShouldBindQuery(&req); err != nil {
+		utils.MakeInvalidRequestErrorFromError(c, err)
+		return
+	}
+	db := utils.GetTiDBConnection(c)
+	fields := []string{}
+	if strings.TrimSpace(req.Fields) != "" {
+		fields = strings.Split(req.Fields, ",")
+	}
+	overviews, err := QueryStatementsOverview(
+		db,
+		req.BeginTime, req.EndTime,
+		req.Schemas,
+		req.StmtTypes,
+		req.Text,
+		fields)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	if len(overviews) == 0 {
+		utils.MakeInvalidRequestErrorFromError(c, errors.New("no data to export"))
+		return
+	}
+
+	// convert data
+	fieldsMap := make(map[string]string)
+	t := reflect.TypeOf(overviews[0])
+	fieldsNum := t.NumField()
+	for i := 0; i < fieldsNum; i++ {
+		field := t.Field(i)
+		fieldsMap[strings.ToLower(field.Tag.Get("json"))] = field.Name
+	}
+
+	csvData := [][]string{fields}
+	for _, overview := range overviews {
+		row := []string{}
+		v := reflect.ValueOf(overview)
+		for _, field := range fields {
+			filedName := fieldsMap[field]
+			s := v.FieldByName(filedName)
+			var val string
+			switch s.Interface().(type) {
+			case int:
+				val = strconv.FormatInt(s.Int(), 10)
+			default:
+				val = s.String()
+			}
+			row = append(row, val)
+		}
+		csvData = append(csvData, row)
+	}
+
+	// generate csv
+	timeLayout := "01021504"
+	beginTime := time.Unix(int64(req.BeginTime), 0).Format(timeLayout)
+	endTime := time.Unix(int64(req.EndTime), 0).Format(timeLayout)
+	tmpfile, err := ioutil.TempFile("", fmt.Sprintf("statements_%s_%s_*.csv", beginTime, endTime))
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	defer tmpfile.Close()
+	csvwriter := csv.NewWriter(tmpfile)
+	for _, csvRow := range csvData {
+		_ = csvwriter.Write(csvRow)
+	}
+	csvwriter.Flush()
+
+	// generate token by filepath
+	token, err := utils.NewJWTString("statements/download", tmpfile.Name())
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.String(http.StatusOK, token)
+}
+
+// @Router /statements/download [get]
+// @Summary Download statements
+// @Produce application/zip
+// @Param token query string true "download token"
+// @Failure 400 {object} utils.APIError
+// @Failure 401 {object} utils.APIError "Unauthorized failure"
+func (s *Service) downloadHandler(c *gin.Context) {
+	token := c.Query("token")
+	fielPath, err := utils.ParseJWTString("statements/download", token)
+	if err != nil {
+		utils.MakeInvalidRequestErrorFromError(c, err)
+		return
+	}
+
+	_, err = os.Stat(fielPath)
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	c.Writer.Header().Set("Content-type", "application/octet-stream")
+	c.Writer.Header().Set("Content-Disposition", "attachment; filename=\"statements.zip\"")
+	err = utils.StreamZipPack(c.Writer, []string{fielPath}, true)
+	if err != nil {
+		log.Error("Stream zip pack failed", zap.Error(err))
+	}
```

Comment:
I wonder how to deal with the temporary .csv files if don't delete them after each download (though this is not a good solution)?

Reply:
Because we put them in the system `/tmp` folder, so they will be cleaned up by the system, for example, they will be deleted automatically when rebooting.

Reply:
Maybe deleting the file after request can be a good idea to minimize the temp file size..

Reply:
Done.